OCPBUGS-7374: set default timeouts in etcdcli

lance5890 opened the pull request:
The ceo operator connect to etcd without timeout, which may lead to some ceo controllers(like EtcdEndpointsController, EtcdCertSignerController and so on ) may be stuck in etcd member list interface about 15minutes, which finally may be closed by the operating system after 15 minutes

[issue1000: fix ceo connection timeout ](https://github.com/openshift/cluster-etcd-operator/issues/1000)

1. this problem can be reproduced by rm one master node, and the pprof top info will show like this:
<img width="672" alt="image" src="https://user-images.githubusercontent.com/30207518/218005561-1792bc6a-547b-42fb-936d-f1d6d6d33e75.png">

2. maybe the 60s is not suitable for most conditions，we can discuss about this

Signed-off-by: lan.tian <lance5890@163.com>

lance5890 commented:
/assign @tjungblu

lance5890 commented:
@dusk125

tjungblu commented:
/ok-to-test

tjungblu commented:
Thanks. @dusk125 is out for the coming weeks, but I'm happy to assist.
60s sounds high for almost all calls, maybe we should reduce this to 30s? Maybe defrag needs a different timeout, that can take much longer.

I do remember I have changed something similar in the past, we discovered controllers getting stuck in some calls before already. I believe I've added a timeout for the respective calls there too. Let me dig this up again.

lance5890 commented:
> 

ok, we can reduce the timeout time by different situation, this PR only care about the connection timeout with the etcd，and we always need client timeout , i will modify the default timeout to 30s

lance5890 commented:
> 

and the defrag controller use the etcd status interface, we can only change the etcd status interface timeout to suitable value

lance5890 commented:
> Thanks. @dusk125 is out for the coming weeks, but I'm happy to assist. 60s sounds high for almost all calls, maybe we should reduce this to 30s? Maybe defrag needs a different timeout, that can take much longer.
> 
> I do remember I have changed something similar in the past, we discovered controllers getting stuck in some calls before already. I believe I've added a timeout for the respective calls there too. Let me dig this up again.

I have reposted my code again, change default timeout to 30s, change etcd status timeout to DefragDialTimeout(60s), and that would be my honor to merge my contribution within ocp

tjungblu commented:
found it, that was in a slightly different place with the same issue: 
https://github.com/openshift/cluster-etcd-operator/commit/4816fab709e11e0681b760003be3f1de12c9c103

I think adding this directly to the client makes sense with 30s as a default. Let me quickly create all the bug tickets around this, I'll carry this through the other releases for you.

tjungblu commented on pkg/cmd/verify/backupstorage.go:
```diff
@@ -132,6 +132,8 @@ func (v *verifyBackupStorage) isStorageAdequate(ctx context.Context) (bool, erro
 
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
 	members, err := cli.MemberList(ctx)
```
why would that listing take 60s?

lance5890 replied:
I will reduce this to 30s

tjungblu commented on pkg/etcdcli/etcdcli.go:
```diff
@@ -39,6 +39,7 @@ const (
 	BootstrapIPAnnotationKey = "alpha.installer.openshift.io/etcd-bootstrap"
 	DefaultDialTimeout       = 15 * time.Second
 	DefragDialTimeout        = 60 * time.Second
+	DefaultCliTimeout        = 30 * time.Second
```
feel free to spell out Client, we're not short on characters today

lance5890 replied:
ok， i will use DefaultClientTimeout

tjungblu commented on pkg/etcdcli/etcdcli.go:
```diff
@@ -282,6 +293,8 @@ func (g *etcdClientGetter) Status(ctx context.Context, clientURL string) (*clien
 	}
 
 	defer g.clientPool.Return(cli)
+	ctx, cancel := context.WithTimeout(ctx, DefragDialTimeout)
```
I think that's fine and can be 30s, the actual defragmentation call could take longer. Can you change that back to `DefaultCliTimeout` please?

lance5890 replied:
ok， i will change it back to DefaultClientTimeout

lance5890 commented:
> found it, that was in a slightly different place with the same issue: [4816fab](https://github.com/openshift/cluster-etcd-operator/commit/4816fab709e11e0681b760003be3f1de12c9c103)
> 
> I think adding this directly to the client makes sense with 30s as a default. Let me quickly create all the bug tickets around this, I'll carry this through the other releases for you.

all done， please review again , thanks!

tjungblu commented:
/retitle OCPBUGS-7374: set default timeouts in etcdcli

tjungblu commented:
awesome, thank you. 

/lgtm

lance5890 commented:
> /cherry-pick release-4.12

4.11 has the same problem, could u cherry-pick to 4.11

tjungblu commented:
yes, I will backport both of your fixes to 4.9.  We have to backport them in order of releases.

tjungblu commented:
/retest

tjungblu commented:
Thanks for your contribution @lance5890, I'll get the backports as soon as possible. It might take some time to end up in an official z-release however.

lance5890 commented:
> Thanks for your contribution @lance5890, I'll get the backports as soon as possible. It might take some time to end up in an official z-release however.

it is my pleasure to cooperate with u, i will keep following with ceo, and i hope to contribute more than these two bugfixs in https://github.com/openshift/cluster-etcd-operator/issues/1000
